### PR TITLE
Generate parser from `cwltool` extensions

### LINKED ISSUE #309: `schema_salad.exceptions.ValidationException` in parser
Given the workflow below:
```cwl
#!/usr/bin/env cwl-runner

cwlVersion: v1.0
class: CommandLineTool
$namespaces:
  cwltool: "http://commonwl.org/cwltool#"

doc: |
  Trivial MPI test that prints the process IDs of each of the parallel
  processes. Requires Python (but you have cwltool running, right?)
  and an MPI implementation.

baseCommand: python
requirements:
  cwltool:MPIRequirement:
    processes: 2
arguments: [-c, 'import os; print(os.getpid())']
inputs: []
outputs:
  pids:
    type: stdout
```

Running `cwl-docker-extract mpi_simple.cwl` crashes at parsing:
```
$ cwl-docker-extract file:///home/heaucques/Documents/toil/src/toil/test/cwl/mpi_simple.cwl
Traceback (most recent call last):
  File "/home/heaucques/Documents/toil/venv3.12/bin/cwl-docker-extract", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/heaucques/Documents/toil/venv3.12/lib/python3.12/site-packages/cwl_utils/docker_extract.py", line 139, in main
    run(arg_parser().parse_args(sys.argv[1:]))
  File "/home/heaucques/Documents/toil/venv3.12/lib/python3.12/site-packages/cwl_utils/docker_extract.py", line 55, in run
    top = cwl.load_document_by_uri(args.input)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/heaucques/Documents/toil/venv3.12/lib/python3.12/site-packages/cwl_utils/parser/__init__.py", line 183, in load_document_by_uri
    return load_document_by_string(doc, real_uri, loadingOptions, id_, load_all)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/heaucques/Documents/toil/venv3.12/lib/python3.12/site-packages/cwl_utils/parser/__init__.py", line 211, in load_document_by_string
    return load_document_by_yaml(result, uri, loadingOptions, id_, load_all)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/heaucques/Documents/toil/venv3.12/lib/python3.12/site-packages/cwl_utils/parser/__init__.py", line 227, in load_document_by_yaml
    result = cwl_v1_0.load_document_by_yaml(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/heaucques/Documents/toil/venv3.12/lib/python3.12/site-packages/cwl_utils/parser/cwl_v1_0.py", line 21054, in load_document_by_yaml
    result, metadata = _document_load(
                       ^^^^^^^^^^^^^^^
  File "/home/heaucques/Documents/toil/venv3.12/lib/python3.12/site-packages/cwl_utils/parser/cwl_v1_0.py", line 1038, in _document_load
    loader.load(doc, baseuri, loadingOptions, docRoot=baseuri),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/heaucques/Documents/toil/venv3.12/lib/python3.12/site-packages/cwl_utils/parser/cwl_v1_0.py", line 806, in load
    raise ValidationException("", None, errors, "*")
schema_salad.exceptions.ValidationException: src/toil/test/cwl/mpi_simple.cwl:3:1:   Object `mpi_simple.cwl` is not valid because:
src/toil/test/cwl/mpi_simple.cwl:14:1:   the `requirements` field is not valid because:
                                           Expected a <class 'NoneType'> but got list
```

### PR DESCRIPTION
**Pull request by GlassOfWhiskey**
This commit fixes #309 by generating the Python CWL parsers directly from the `codegen` branch of cwltool, including all the cwltool extensions in the parsers logic.

### CODE REVIEW

**Review by mr-c — CHANGES_REQUESTED** [review]
Thanks for this! Can you open PRs to the codegen branches of https://github.com/common-workflow-language/cwl-v1.2 etc. to add the extensions there?

**Comment by GlassOfWhiskey** [comment]
> Thanks for this! Can you open PRs to the codegen branches of https://github.com/common-workflow-language/cwl-v1.2 etc. to add the extensions there?

Hi @mr-c, I just did it. Would you like to release `schema_salad`, in order to remove the git dependency?

File: README.rst [hunk 1]
```diff
@@ -152,13 +152,13 @@ Regenerate parsers
...
+``schema-salad-tool --codegen python https://github.com/common-workflow-language/common-workflow-language/raw/codegen/v1.0/extensions.yml --codegen-parser-info "org.w3id.cwl.v1_0" > cwl_utils/parser/cwl_v1_0.py``
 
 ``cwl_utils/parser/cwl_v1_1.py`` was created via
-``schema-salad-tool --codegen python https://github.com/common-workflow-language/cwl-v1.1/raw/codegen/CommonWorkflowLanguage.yml --codegen-parser-info "org.w3id.cwl.v1_1" > cwl_utils/parser/cwl_v1_1.py``
+``schema-salad-tool --codegen python https://github.com/common-workflow-language/cwl-v1.1/raw/codegen/extensions.yml --codegen-parser-info "org.w3id.cwl.v1_1" > cwl_utils/parser/cwl_v1_1.py``
 
 ``cwl_utils/parser/cwl_v1_2.py`` was created via
-``schema-salad-tool --codegen python https://github.com/common-workflow-language/cwl-v1.2/raw/1.2.1_proposed/CommonWorkflowLanguage.yml --codegen-parser-info "org.w3id.cwl.v1_2" > cwl_utils/parser/cwl_v1_2.py``
+``schema-salad-tool --codegen python https://github.com/common-workflow-language/cwl-v1.2/raw/codegen/extensions.yml --codegen-parser-info "org.w3id.cwl.v1_2" > cwl_utils/parser/cwl_v1_2.py``
```
Comment (mr-c) [inline README.rst:161]: This should be updated in the `Makefile` as well

File: requirements.txt [hunk 2]
```diff
@@ -2,5 +2,5 @@ cwl-upgrader >= 1.2.3
 packaging
 rdflib
 requests
-schema-salad >= 8.8, < 9
+schema-salad @ git+https://github.com/common-workflow-language/schema_salad.git
```
Comment (mr-c) [inline requirements.txt:5]: Is this still needed?

**Review by mr-c — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 2 above)
Comment (GlassOfWhiskey) [reply requirements.txt]: No, but we need `>= 8.8.20241206093842` for the proper detection of namespaced classes

Comment (GlassOfWhiskey) [reply requirements.txt]: Oh no that's not true. This is true just during generation, not at runtime.

File: README.rst (hunk 1 above)
Comment (GlassOfWhiskey) [reply README.rst:161]: But which `schema_salad` version is picked by the makefile?

Comment (GlassOfWhiskey) [reply README.rst:161]: If Makefile uses the `schema_salad` from `requirements.txt`, then it should be `schema_salad >= 8.8.20241206093842,<9`

**Review by mr-c — APPROVED** [review]

**Pull request merged**